scrapy/settings/__init__.py

```diff
@@ -233,6 +235,47 @@ def maxpriority(self):
         else:
             return get_settings_priority('default')
 
+    def getinstance(self, name, *, crawler=None, singleton=False):
```

Comment:
Aha, I see the problem now: you can't just create RequestFingerprinter in a dupefilter, because other components also may need to access it.

I'm not sure though that settings is a best place to handle this singleton. 

1) All other settings return simple values at the moment - str, list, dict.
2) You need to pass crawler here, but only once - on a first call. On all other calls crawler is not necessary. But we need to ensure that the first call was a call which passed a crawler.

I haven't checked it in detail, but what do you think about making fingerprinter a more "fundamental" thing, maybe a Crawler attribute, similar to logformatter or stats? In this case logic around singletons won't be needed.

Reply:
This may require an access to Crawler in some places where it was not possible before though; it could make the scope of this PR larger.